File: Data Structures & Algorithms/islands-and-treasure/submission-2.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:

        self.ROWS, self.COLS = len(grid), len(grid[0])
        self.INF = 2147483647

        q = collections.deque()
        for r in range(self.ROWS):
            for c in range(self.COLS):
                if grid[r][c] == 0:
                    q.append((r, c))

        directions = [[-1, 0], [+1, 0], [0, -1], [0, +1]]
        visited = set()
        while q:

            for _ in range(len(q)):

                r, c = q.popleft()
                
                visited.add((r, c))
                for dir in directions:
                    nr, nc = r + dir[0], c + dir[1]
                    if min(nr, nc) >= 0 and nr < self.ROWS and nc < self.COLS and (nr, nc) not in visited and grid[nr][nc] == self.INF:
                        grid[nr][nc] = grid[r][c] + 1
                        q.append((nr, nc))

File: Data Structures & Algorithms/matrixBFS/submission-2.cpp
class Solution {
public:
    int shortestPath(vector<vector<int>>& grid) {
        int ROWS = grid.size(), COLS = grid[0].size();

        vector<vector<bool>> visited(ROWS, vector<bool>(COLS, false));
        queue<vector<int>> q;

        if (!grid[0][0]){
            q.push({0, 0});
        }

        int minLen = 0;
        
        while (!q.empty()){
            int size = q.size();
            for (int i = 0; i < size; i++){
                int r = q.front()[0], c = q.front()[1];

                q.pop();

                if (r == ROWS - 1 && c == COLS - 1){
                    return minLen;
                }

                int neighbors[4][2] = {{r - 1, c}, {r + 1, c}, {r, c - 1}, {r, c + 1}};

                for (int j = 0; j < 4; j++){
                    int rnei = neighbors[j][0], cnei = neighbors[j][1];

                    if (min(rnei, cnei) < 0 || rnei >= ROWS || cnei >= COLS || grid[rnei][cnei] || visited[r][c]){
                        continue;
                    }
                    q.push({rnei, cnei});
                }
                visited[r][c] = true;
            }
            minLen++;
        }

        return -1;
    }
};
